### DISCUSSION TITLE
Intercepting messages

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [nasa/cFS#612] [General]
Hi,

I'm trying to implement a fault-injection capability that will intercept and modify cFS messages, for V&V purposes.

Do you know if stopping messages from being sent to other applications is possible? What would be the best way to approach this?

Thanks

**Comment by skliper** [comment]
Brainstormed, "risky" way to do it and untested but should work in theory:

Make a fault injection task a higher priority than the task under test.  Subscribe to the same messages, and modify the data in those messages (which you should never do in production).  By the time the task under test runs it'll end up processing the modified message.

**Comment by ivanperez-keera** [comment]
Thanks.

Fault injection will only run during tests, so this could be alright.

By modify the messages, do you mean take the message content, cast it to the right struct type, then just modify the struct fields and that should be enough? Or is there a different mechanism to overwrite messages?

Does everything run in the same thread? Is there a guarantee that the routine that modifies the messages will finish before any lower-priority task receives the (still unmodified) message?

**Reply by skliper** [reply]
A message sits in shared memory space owned by the software bus and the pointer is just provided to all receivers of the message.  From the fault injection side, subscribe to the message, do a receive, cast based on ID and modify.  Threading and guarantees all depend on how you've set up your system on your platform.  If you are allowed to modify message ID's you could just have a pre and post corrupted ID then the low level details wouldn't really matter (and you wouldn't have to modify a message already on the software bus).
